source/base/parameter_acceptor.cc

```diff
@@ -24,25 +24,48 @@
 DEAL_II_NAMESPACE_OPEN
 
 
-// Static empty class list
-std::vector<SmartPointer<ParameterAcceptor>> ParameterAcceptor::class_list;
+namespace internal
+{
+  struct ParameterAcceptorCompare
+  {
+    bool
+    operator()(const ParameterAcceptor *p1, const ParameterAcceptor *p2) const
+    {
+      return p1->get_acceptor_id() < p2->get_acceptor_id();
+    }
+  };
+} // namespace internal
+
+
+// Static mutex for the class list
+std::mutex ParameterAcceptor::class_list_mutex;
+
+// Static empty class set
+std::set<ParameterAcceptor *, internal::ParameterAcceptorCompare>
+  ParameterAcceptor::class_list;
```

Comment:
I think the forward declaration works, because we first instantiate this here, and everything is known at this point.

Reply:
Actually, upon second though, it is because `std::` containers do not actually store objects of this type -- they're stateless, and they're instantiated every time some place needs such an object.